lib/types.nix

```diff
@@ -710,6 +710,23 @@ rec {
         in mergedOption.type;
     };
 
+    # A type that is one of several submodules, similiar to types.oneOf but is usable inside attrsOf or listOf
+    # submodules need an option with a type str which is used to find the corresponding type
+    taggedSubmodules =
+      { types
+      , specialArgs ? {}
+      }: mkOptionType rec {
+      name = "taggedSubmodules";
+      description = "one of ${concatStringsSep "," (attrNames types)}";
+      check = x: if x ? type then types.${x.type}.check x else throw "No type option set in:\n${lib.generators.toPretty {} x}";
+      merge = loc: foldl'
+        (res: def: types.${def.value.type}.merge loc [
+          (lib.recursiveUpdate { value._module.args = specialArgs; } def)
+        ])
```

Comment:
This makes it impossible btw to use `mkMerge` if one of the items to be merged is missing a `type`. Minimal example in `disko` where I encountered this ~yesterday:

``` nix
{ lib, ... }:

with lib;

let
  mkDataset = cfg: mkMerge [ cfg { type = "zfs_fs"; } ];
  mkLegacyMount = p: { mountpoint = p; options.mountpoint = "none"; };
in {
  disko.devices.zpool.tank.datasets = mapAttrs (const mkDataset) {
    "foo" = mkLegacyMount "/foo";
    "bar" =mkLegacyMount "/bar";
  };
}
```

This gives the following error:

```
       error: No type option set in:
       {
         mountpoint = "/bar";
         options = {
           mountpoint = "none";
         };
       }
```

AFAIU this happens because each declaration on its own is passed into the wrapped type, i.e. the stuff from `mkLegacyMount` is passed unmerged, i.e. without the `type` from `mkDataset`.

Anyways, having attr-sets inside the module system where the surroundings (i.e. `mkIf`/`mkMerge`/etc) "silently" stop working[1] isn't very nice and it would be cool if this could get fixed. Alternatively I'd also be fine with a better error message that catches this issue properly (it's already rather simple to break a module evaluation in a way that it's hard to figure out the culprit and I'd rather not have yet another case).

[1] well, it fails, but it doesn't tell you that it fails because of `mkMerge`.

Reply:
This could probably be solved by doing an evalModules just to figure out the type and then evaluate the submodule again. (See https://github.com/NixOS/nixpkgs/pull/254790/files#r1323465281)